File: cpp/src/puct.cpp
#include "hex_puct/puct.hpp"

#include <algorithm>
#include <cmath>
#include <queue>
#include <stdexcept>

namespace hex_puct {

Color Opponent(Color color) { return color == Color::Black ? Color::White : Color::Black; }

Board::Board(Color side_to_move) : side_to_move_(side_to_move) {}

Board Board::FromSetup(const std::vector<int>& black, const std::vector<int>& white,
                       Color side_to_move, std::optional<int> last_move) {
  Board board(side_to_move);
  for (int a : black) board.Place(a, Color::Black);
  for (int a : white) board.Place(a, Color::White);
  if (last_move && (*last_move < 0 || *last_move >= kBoardArea || board.cells_[*last_move] == 0))
    throw std::invalid_argument("invalid last move");
  board.last_move_ = last_move;
  return board;
}

void Board::Place(int action, Color color) {
  if (action < 0 || action >= kBoardArea || cells_[action]) throw std::invalid_argument("invalid setup action");
  cells_[action] = color == Color::Black ? 1 : 2;
}

void Board::Play(int action) {
  if (LiteralWinner() || action < 0 || action >= kBoardArea || cells_[action])
    throw std::invalid_argument("illegal action");
  Place(action, side_to_move_);
  last_move_ = action;
  side_to_move_ = Opponent(side_to_move_);
}

std::vector<int> Board::LegalActions() const {
  if (LiteralWinner()) return {};
  std::vector<int> result;
  for (int a = 0; a < kBoardArea; ++a) if (!cells_[a]) result.push_back(a);
  return result;
}

std::array<bool, kBoardArea> Board::LegalMask() const {
  std::array<bool, kBoardArea> out{};
  if (!LiteralWinner()) for (int a = 0; a < kBoardArea; ++a) out[a] = !cells_[a];
  return out;
}

std::optional<Color> Board::LiteralWinner() const {
  constexpr int dx[6] = {-1, -1, 0, 0, 1, 1};
  constexpr int dy[6] = {0, 1, -1, 1, -1, 0};
  for (Color c : {Color::Black, Color::White}) {
    unsigned char stone = c == Color::Black ? 1 : 2;
    std::queue<int> q;
    std::array<bool, kBoardArea> seen{};
    for (int a = 0; a < kBoardArea; ++a) {
      int x = a % kBoardSize, y = a / kBoardSize;
      if (cells_[a] == stone && ((c == Color::Black && y == 0) || (c == Color::White && x == 0))) {
        q.push(a); seen[a] = true;
      }
    }
    while (!q.empty()) {
      int a = q.front(); q.pop();
      int x = a % kBoardSize, y = a / kBoardSize;
      if ((c == Color::Black && y == kBoardSize - 1) || (c == Color::White && x == kBoardSize - 1)) return c;
      for (int i = 0; i < 6; ++i) {
        int nx = x + dx[i], ny = y + dy[i];
        if (nx >= 0 && nx < kBoardSize && ny >= 0 && ny < kBoardSize) {
          int n = ny * kBoardSize + nx;
          if (!seen[n] && cells_[n] == stone) { seen[n] = true; q.push(n); }
        }
      }
    }
  }
  return std::nullopt;
}

int Board::Signature() const {
  int s = side_to_move_ == Color::Black ? 17 : 29;
  for (int a = 0; a < kBoardArea; ++a) s = (s + (a + 1) * cells_[a]) % 1000003;
  return s;
}

double Edge::Q() const { return visits ? value_sum / visits : 0.0; }

double ComputeFpuQ(FpuMode mode, double parent_value, double reduction, double visited_prior_mass) {
  if (mode == FpuMode::Zero) return 0.0;
  const double raw = parent_value - reduction * std::sqrt(std::max(0.0, visited_prior_mass));
  return std::max(-1.0, std::min(1.0, raw));
}

DeterministicPUCT::DeterministicPUCT(Evaluator evaluator, SearchConfig config)
    : evaluator_(std::move(evaluator)), config_(config) {
  if (config.simulations < 0 || !std::isfinite(config.c_puct) || config.c_puct < 0 ||
      !std::isfinite(config.fpu_reduction) || config.fpu_reduction < 0)
    throw std::invalid_argument("invalid config");
  if (config.fpu_mode == FpuMode::Zero && config.fpu_reduction != 0)
    throw std::invalid_argument("zero FPU cannot have a reduction");
}

double DeterministicPUCT::Expand(Node& node, const Board& board) const {
  auto e = evaluator_(board);
  if (!std::isfinite(e.value) || e.value < -1.000001 || e.value > 1.000001)
    throw std::invalid_argument("invalid value");
  node.expansion_value = e.value;
  auto legal = board.LegalActions();
  if (legal.empty()) return e.value;
  double maximum = -INFINITY;
  for (int a : legal) {
    if (!std::isfinite(e.policy_logits[a])) throw std::invalid_argument("invalid legal logit");
    maximum = std::max(maximum, e.policy_logits[a]);
  }
  double total = 0;
  for (int a : legal) total += std::exp(e.policy_logits[a] - maximum);
  if (!std::isfinite(total) || total <= 0) throw std::invalid_argument("invalid priors");
  node.edges.clear(); node.edges.reserve(legal.size());
  for (int a : legal) node.edges.push_back({a, std::exp(e.policy_logits[a] - maximum) / total});
  return e.value;
}

double DeterministicPUCT::FpuQ(const Node& node) const {
  if (config_.fpu_mode == FpuMode::Zero) return 0.0;
  double visited_prior = 0.0;
  for (const auto& edge : node.edges) if (edge.visits > 0) visited_prior += edge.prior;
  return ComputeFpuQ(config_.fpu_mode, node.expansion_value, config_.fpu_reduction, visited_prior);
}

Edge& DeterministicPUCT::Select(Node& node) const {
  double scale = config_.c_puct * std::sqrt(std::max(1, node.visits));
  const double fpu = FpuQ(node);
  Edge* best = nullptr; double best_score = -INFINITY;
  for (auto& e : node.edges) {
    double score = (e.visits ? e.Q() : fpu) + scale * e.prior / (1 + e.visits);
    if (!best || score > best_score || (score == best_score && e.action < best->action)) {
      best = &e; best_score = score;
    }
  }
  return *best;
}

SearchResult DeterministicPUCT::Search(const Board& source) const {
  SearchResult r;
  if (auto winner = source.LiteralWinner()) {
    r.root_value = *winner == source.side_to_move() ? 1.0 : -1.0;
    return r;
  }
  Node root{source.side_to_move()};
  Expand(root, source); r.evaluations = 1;
  for (int i = 0; i < config_.simulations; ++i) {
    Board current = source; Node* node = &root;
    std::vector<Node*> nodes{&root}; std::vector<Edge*> edges;
    while (!node->edges.empty() && !current.LiteralWinner()) {
      Edge& edge = Select(*node); current.Play(edge.action); edges.push_back(&edge);
      if (!edge.child) edge.child = std::make_unique<Node>(Node{current.side_to_move()});
      node = edge.child.get(); nodes.push_back(node);
      if (node->visits == 0 || node->edges.empty()) break;
    }
    r.max_depth = std::max(r.max_depth, static_cast<int>(edges.size()));
    double leaf;
    if (auto winner = current.LiteralWinner()) leaf = *winner == current.side_to_move() ? 1.0 : -1.0;
    else { leaf = Expand(*node, current); ++r.evaluations; }
    for (Node* n : nodes) ++n->visits;
    for (auto it = edges.rbegin(); it != edges.rend(); ++it) { leaf = -leaf; ++(*it)->visits; (*it)->value_sum += leaf; }
  }
  int maximum_visits = -1;
  for (const auto& e : root.edges) {
    r.raw_visits[e.action] = e.visits; r.priors[e.action] = e.prior;
    r.raw_value_sums[e.action] = e.value_sum; r.root_visits += e.visits;
    if (e.visits > maximum_visits) { maximum_visits = e.visits; r.selected_action = e.action; }
  }
  for (int a = 0; a < kBoardArea; ++a) r.policy[a] = r.root_visits ? static_cast<double>(r.raw_visits[a]) / r.root_visits : 0;
  if (r.root_visits) { double sum = 0; for (const auto& e : root.edges) sum += e.value_sum; r.root_value = sum / r.root_visits; }
  return r;
}

Evaluation DeterministicFakeEvaluator::operator()(const Board& board) const {
  Evaluation e{}; int s = board.Signature();
  for (int a = 0; a < kBoardArea; ++a) e.policy_logits[a] = ((a * 17 + s * 31) % 37 - 18) / 8.0;
  e.value = ((s % 13) - 6) / 7.0; return e;
}

}  // namespace hex_puct

File: src/hex_reconstruction/puct.py
"""Deterministic single-threaded PUCT reference search for physical 11x11 Hex.

The evaluator value is always from the *side-to-move in the evaluated board*
perspective.  Edge values are stored from their parent state's side-to-move
perspective.  That convention makes alternating-sign backup explicit.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from pathlib import Path
from typing import Protocol, Sequence

from .board import BOARD_AREA, HexBoard, opponent


@dataclass(frozen=True)
class Evaluation:
    """Raw 121 physical-action logits and side-to-move scalar value in [-1, 1]."""

    policy_logits: Sequence[float]
    value: float


class Evaluator(Protocol):
    def evaluate(self, position: HexBoard) -> Evaluation: ...


@dataclass
class Edge:
    action: int
    prior: float
    child: "Node | None" = None
    visits: int = 0
    value_sum: float = 0.0

    @property
    def q(self) -> float:
        return self.value_sum / self.visits if self.visits else 0.0


@dataclass
class Node:
    """A state node. `visits` is the number of completed simulations through it."""

    to_play: str
    visits: int = 0
    edges: list[Edge] = field(default_factory=list)
    expansion_value: float = 0.0


@dataclass(frozen=True)
class SearchConfig:
    simulations: int
    c_puct: float = 1.5
    fpu_mode: str = "zero"
    fpu_reduction: float = 0.0

    def __post_init__(self) -> None:
        if self.simulations < 0:
            raise ValueError("simulations must be nonnegative")
        if not math.isfinite(self.c_puct) or self.c_puct < 0:
            raise ValueError("c_puct must be nonnegative")
        if self.fpu_mode not in {"zero", "parent_value_reduced"}:
            raise ValueError("unsupported fpu_mode")
        if not math.isfinite(self.fpu_reduction) or self.fpu_reduction < 0 or (self.fpu_mode == "zero" and self.fpu_reduction != 0):
            raise ValueError("invalid fpu_reduction")


@dataclass(frozen=True)
class SearchResult:
    selected_action: int | None
    root_visits: int
    raw_visits: tuple[int, ...]
    policy: tuple[float, ...]
    root_value: float | None
    priors: tuple[float, ...]
    raw_value_sums: tuple[float, ...] = ()
    evaluations: int = 0
    max_depth: int = 0


def compute_fpu_q(mode: str, parent_value: float, reduction: float, visited_prior_mass: float) -> float:
    if mode == "zero":
        return 0.0
    if mode != "parent_value_reduced" or reduction < 0:
        raise ValueError("invalid FPU configuration")
    raw = parent_value - reduction * math.sqrt(max(0.0, visited_prior_mass))
    return max(-1.0, min(1.0, raw))


def _terminal_value(position: HexBoard) -> float:
    """Value from `position.side_to_move`; legal Hex terminal states normally give -1."""

    winner = position.literal_winner()
    if winner is None:
        raise ValueError("terminal value requested for non-terminal position")
    return 1.0 if winner == position.side_to_move else -1.0


def _legal_priors(position: HexBoard, logits: Sequence[float]) -> list[Edge]:
    if len(logits) != BOARD_AREA:
        raise ValueError(f"evaluator must return {BOARD_AREA} logits")
    legal = position.legal_actions()
    if not legal:
        return []
    selected = [float(logits[action]) for action in legal]
    if not all(math.isfinite(value) for value in selected):
        raise ValueError("evaluator logits for legal actions must be finite")
    maximum = max(selected)
    weights = [math.exp(value - maximum) for value in selected]
    total = sum(weights)
    if not math.isfinite(total) or total <= 0:
        raise ValueError("unable to normalize legal policy priors")
    return [Edge(action=action, prior=weight / total) for action, weight in zip(legal, weights)]


class DeterministicPUCT:
    """Reference PUCT: one thread, no virtual loss, no pruning, no batching.

    Selection score is ``Q + c_puct * P * sqrt(max(1, N_parent)) / (1+N_edge)``.
    Equal scores are resolved by smallest physical action index.  Root action
    selection is maximum raw edge visits, then smallest physical action index.
    """

    def __init__(self, evaluator: Evaluator, config: SearchConfig) -> None:
        self.evaluator = evaluator
        self.config = config

    def _expand(self, node: Node, position: HexBoard) -> float:
        evaluation = self.evaluator.evaluate(position)
        if not math.isfinite(evaluation.value) or not -1.000001 <= evaluation.value <= 1.000001:
            raise ValueError("evaluator value must be finite and within [-1,1]")
        node.expansion_value = float(evaluation.value)
        node.edges = _legal_priors(position, evaluation.policy_logits)
        return float(evaluation.value)

    def _fpu_q(self, node: Node) -> float:
        if self.config.fpu_mode == "zero":
            return 0.0
        visited_prior = sum(edge.prior for edge in node.edges if edge.visits > 0)
        return compute_fpu_q(self.config.fpu_mode, node.expansion_value, self.config.fpu_reduction, visited_prior)

    def _select(self, node: Node) -> Edge:
        if not node.edges:
            raise RuntimeError("cannot select from unexpanded node")
        scale = self.config.c_puct * math.sqrt(max(1, node.visits))
        fpu = self._fpu_q(node)
        return max(node.edges, key=lambda edge: ((edge.q if edge.visits else fpu) + scale * edge.prior / (1 + edge.visits), -edge.action))

    def search(self, source: HexBoard) -> SearchResult:
        """Search a copy; `source` is never changed."""

        position = source.copy()
        if position.literal_winner() is not None:
            zero = tuple(0 for _ in range(BOARD_AREA))
            return SearchResult(None, 0, zero, tuple(0.0 for _ in zero), _terminal_value(position), tuple(0.0 for _ in zero), tuple(0.0 for _ in zero))

        root = Node(to_play=position.side_to_move)
        self._expand(root, position)
        evaluations = 1
        max_depth = 0
        for _ in range(self.config.simulations):
            current = source.copy()
            node = root
            node_path = [root]
            edge_path: list[Edge] = []

            while node.edges and current.literal_winner() is None:
                edge = self._select(node)
                current.play(edge.action)
                edge_path.append(edge)
                if edge.child is None:
                    edge.child = Node(to_play=current.side_to_move)
                node = edge.child
                node_path.append(node)
                if node.visits == 0 or not node.edges:
                    break

            if current.literal_winner() is not None:
                leaf_value = _terminal_value(current)
            else:
                leaf_value = self._expand(node, current)
                evaluations += 1
            max_depth = max(max_depth, len(edge_path))

            # `leaf_value` is for current.to_play. Each parent is its opponent.
            for path_node in node_path:
                path_node.visits += 1
            value = leaf_value
            for edge in reversed(edge_path):
                value = -value
                edge.visits += 1
                edge.value_sum += value

        raw = [0] * BOARD_AREA
        priors = [0.0] * BOARD_AREA
        raw_value_sums = [0.0] * BOARD_AREA
        for edge in root.edges:
            raw[edge.action] = edge.visits
            priors[edge.action] = edge.prior
            raw_value_sums[edge.action] = edge.value_sum
        total = sum(raw)
        policy = [count / total if total else 0.0 for count in raw]
        selected = min((edge.action for edge in root.edges if edge.visits == max((candidate.visits for candidate in root.edges), default=0)), default=None)
        root_value = sum(edge.value_sum for edge in root.edges) / total if total else None
        return SearchResult(selected, total, tuple(raw), tuple(policy), root_value, tuple(priors), tuple(raw_value_sums), evaluations, max_depth)


class TorchStudentEvaluator:
    """Direct PyTorch bridge; avoids an unqualified historical ONNX/ORT boundary."""

    def __init__(self, checkpoint: Path, *, device: str = "cpu") -> None:
        import torch
        from .student_training import Group49Student

        payload = torch.load(checkpoint, map_location=device, weights_only=False)
        architecture = payload["config"]["architecture"]
        self.model = Group49Student(channels=architecture["channels"], blocks=architecture["residual_blocks"]).to(device)
        self.model.load_state_dict(payload["model_state"])
        self.model.eval()
        self.device = torch.device(device)
        self.checkpoint = Path(checkpoint)
        self.epoch = int(payload.get("epoch", -1))

    def evaluate(self, position: HexBoard) -> Evaluation:
        import torch

        state = torch.tensor(position.feature_planes(), dtype=torch.float32, device=self.device).reshape(1, 6, 11, 11)
        with torch.no_grad():
            logits, value = self.model(state)
        return Evaluation(tuple(float(item) for item in logits[0].detach().cpu()), float(value[0].item()))


class OutputAblationEvaluator:
    """Diagnostic output-level policy/value source mixer.

    It never combines network activations: the full 121 logits come from one
    qualified evaluator and the scalar side-to-move value from another.
    """

    def __init__(self, policy_evaluator: Evaluator, value_evaluator: Evaluator, *, label: str) -> None:
        self.policy_evaluator = policy_evaluator
        self.value_evaluator = value_evaluator
        self.label = label

    def evaluate(self, position: HexBoard) -> Evaluation:
        policy = self.policy_evaluator.evaluate(position)
        value = self.value_evaluator.evaluate(position)
        return Evaluation(policy.policy_logits, value.value)
